From ca0b45bc126ddcc9fefa46a1d160609e1b8dee6b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C3=86var=20Arnfj=C3=B6r=C3=B0=20Bjarmason?=
 <avarab@gmail.com>
Date: Sun, 7 Mar 2021 14:03:29 +0100
Subject: [PATCH 13/16] A bunch of tests fail without the obj->type != OBJ_NONE
 in commit.c

now works:

prove -j8 t6102-rev-list-unexpected-objects.sh t5540-http-push-webdav.sh t5551-http-fetch-smart.sh t5562-http-backend-content-length.sh t5550-http-fetch-dumb.sh t5530-upload-pack-error.sh t5700-protocol-v1.sh t5539-fetch-http-shallow.sh t1450-fsck.sh t*object*.sh
---
 commit.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/commit.c b/commit.c
index 62334dfb0b3..4a5a5d7dedd 100644
--- a/commit.c
+++ b/commit.c
@@ -63,7 +63,12 @@ struct commit *lookup_commit_type(struct repository *r, const struct object_id *
 	struct object *obj = lookup_object(r, oid);
 	if (!obj)
 		return create_object(r, oid, alloc_commit_node(r));
-	if (type != OBJ_NONE)
+	/*
+	 * The obj->type != OBJ_NONE special-case is only for
+	 * lookup_commit(), not lookup_{blob,tree,tag}(). See the
+	 * "obj->type == OBJ_NONE" case in object_as_type().
+	 */
+	if (type != OBJ_NONE && obj->type != OBJ_NONE) 
 		lookup_type_is_type_or_die(oid, obj->type, OBJ_COMMIT);
 	return object_as_type(obj, OBJ_COMMIT, 0);
 }
-- 
2.31.0.rc0.126.g04f22c5b82

